File: app/schemas/timetable.py
from __future__ import annotations

from datetime import date, datetime, time
from enum import Enum

from pydantic import BaseModel, ConfigDict, Field


class TimetableDay(str, Enum):
    MONDAY = "monday"
    TUESDAY = "tuesday"
    WEDNESDAY = "wednesday"
    THURSDAY = "thursday"
    FRIDAY = "friday"
    SATURDAY = "saturday"
    SUNDAY = "sunday"


class TimetableAssignmentType(str, Enum):
    STUDENT = "student"
    TEACHER = "teacher"
    CLASS_GROUP = "class_group"


# =========================================================
# TIMETABLE PERIODS
# =========================================================


class TimetablePeriodBase(BaseModel):
    school_id: int
    name: str = Field(..., max_length=100)
    short_name: str = Field(..., max_length=20)
    period_number: int
    start_time: time
    end_time: time

    is_registration: bool = False
    is_break: bool = False
    is_lunch: bool = False
    is_active: bool = True


class TimetablePeriodCreate(TimetablePeriodBase):
    pass


class TimetablePeriodUpdate(BaseModel):
    name: str | None = None
    short_name: str | None = None
    period_number: int | None = None
    start_time: time | None = None
    end_time: time | None = None

    is_registration: bool | None = None
    is_break: bool | None = None
    is_lunch: bool | None = None
    is_active: bool | None = None


class TimetablePeriodOut(TimetablePeriodBase):
    id: int

    model_config = ConfigDict(from_attributes=True)


# =========================================================
# TIMETABLES
# =========================================================


class TimetableBase(BaseModel):
    school_id: int
    name: str = Field(..., max_length=150)
    academic_year: str = Field(..., max_length=20)

    effective_from: date
    effective_to: date | None = None

    is_active: bool = True


class TimetableCreate(TimetableBase):
    pass


class TimetableUpdate(BaseModel):
    name: str | None = None
    academic_year: str | None = None

    effective_from: date | None = None
    effective_to: date | None = None

    is_active: bool | None = None


class TimetableOut(TimetableBase):
    id: int
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# =========================================================
# TIMETABLE ENTRIES
# =========================================================


class TimetableEntryBase(BaseModel):
    timetable_id: int
    school_id: int

    class_group_id: int | None = None
    course_id: int | None = None
    teacher_id: int | None = None

    timetable_period_id: int

    day_of_week: TimetableDay

    room: str | None = Field(default=None, max_length=100)
    title: str | None = Field(default=None, max_length=200)
    notes: str | None = None


class TimetableEntryCreate(TimetableEntryBase):
    pass


class TimetableEntryUpdate(BaseModel):
    class_group_id: int | None = None
    course_id: int | None = None
    teacher_id: int | None = None

    timetable_period_id: int | None = None

    day_of_week: TimetableDay | None = None

    room: str | None = None
    title: str | None = None
    notes: str | None = None


class TimetableEntryOut(TimetableEntryBase):
    id: int
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# =========================================================
# TIMETABLE ASSIGNMENTS
# =========================================================


class TimetableAssignmentBase(BaseModel):
    timetable_id: int
    school_id: int

    assignment_type: TimetableAssignmentType

    user_id: int | None = None
    class_group_id: int | None = None


class TimetableAssignmentCreate(TimetableAssignmentBase):
    pass


class TimetableAssignmentUpdate(BaseModel):
    user_id: int | None = None
    class_group_id: int | None = None


class TimetableAssignmentOut(TimetableAssignmentBase):
    id: int
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# =========================================================
# FILTERS
# =========================================================


class TimetableFilter(BaseModel):
    school_id: int | None = None
    academic_year: str | None = None
    is_active: bool | None = None

    limit: int = 50
    offset: int = 0


class TimetableEntryFilter(BaseModel):
    school_id: int | None = None

    timetable_id: int | None = None
    class_group_id: int | None = None
    course_id: int | None = None
    teacher_id: int | None = None

    day_of_week: TimetableDay | None = None

    limit: int = 50
    offset: int = 0


class TimetableAssignmentFilter(BaseModel):
    school_id: int | None = None

    timetable_id: int | None = None
    assignment_type: TimetableAssignmentType | None = None

    user_id: int | None = None
    class_group_id: int | None = None

    limit: int = 50
    offset: int = 0

File: mhike-school-web/lib/services/timetable.ts
import { apiGet, apiPost } from "@/lib/api";

export interface TimetablePeriod {
    id: number;
    school_id: number;
    name: string;
    short_name: string;
    period_number: number;
    start_time: string;
    end_time: string;
}

export interface Timetable {
    id: number;
    school_id: number;
    name: string;
    academic_year: string;
    effective_from: string;
    effective_to?: string | null;
    is_active: boolean;
}

export interface TimetableEntry {
    id: number;
    timetable_id: number;
    school_id: number;
    class_group_id: number;
    course_id?: number | null;
    teacher_id?: number | null;
    timetable_period_id: number;
    day_of_week: string;
    title: string;
    room?: string | null;
    notes?: string | null;
}

export interface TimetableAssignment {
    id: number;
    timetable_id: number;
    school_id: number;
    assignment_type: string;
    user_id?: number | null;
    class_group_id?: number | null;
}

export interface TimetableEntryFilters {
    timetable_id?: number;
    class_group_id?: number;
    course_id?: number;
    teacher_id?: number;
    day_of_week?: string;
    limit?: number;
    offset?: number;
}

export interface TimetableFilters {
    academic_year?: string;
    is_active?: boolean;
    limit?: number;
    offset?: number;
}

export interface TimetableAssignmentFilters {
    timetable_id?: number;
    assignment_type?: string;
    user_id?: number;
    class_group_id?: number;
    limit?: number;
    offset?: number;
}

const TIMETABLE_BASE = "/timetables";

function withQuery<T extends object>(
    path: string,
    params?: T,
): string {
    if (!params) {
        return path;
    }

    const searchParams = new URLSearchParams();

    Object.entries(params).forEach(([key, value]) => {
        if (
            value !== undefined &&
            value !== null &&
            value !== ""
        ) {
            searchParams.set(
                key,
                String(value),
            );
        }
    });

    const query = searchParams.toString();

    return query
        ? `${path}?${query}`
        : path;
}

export async function listTimetablePeriods(): Promise<TimetablePeriod[]> {
    return apiGet<TimetablePeriod[]>(
        `${TIMETABLE_BASE}/periods`,
    );
}

export async function createTimetablePeriod(
    payload: Partial<TimetablePeriod>,
): Promise<TimetablePeriod> {
    return apiPost<TimetablePeriod>(
        `${TIMETABLE_BASE}/periods`,
        payload,
    );
}

export async function listTimetables(
    filters?: TimetableFilters,
): Promise<Timetable[]> {
    return apiGet<Timetable[]>(
        withQuery(
            TIMETABLE_BASE,
            filters,
        ),
    );
}

export async function createTimetable(
    payload: Partial<Timetable>,
): Promise<Timetable> {
    return apiPost<Timetable>(
        TIMETABLE_BASE,
        payload,
    );
}

export async function listTimetableEntries(
    filters?: TimetableEntryFilters,
): Promise<TimetableEntry[]> {
    return apiGet<TimetableEntry[]>(
        withQuery(
            `${TIMETABLE_BASE}/entries`,
            filters,
        ),
    );
}

export async function createTimetableEntry(
    payload: Partial<TimetableEntry>,
): Promise<TimetableEntry> {
    return apiPost<TimetableEntry>(
        `${TIMETABLE_BASE}/entries`,
        payload,
    );
}

export async function getTeacherTimetable(
    day_of_week?: string,
): Promise<TimetableEntry[]> {
    return apiGet<TimetableEntry[]>(
        withQuery(
            `${TIMETABLE_BASE}/teacher/me`,
            {
                day_of_week,
            },
        ),
    );
}

export async function getStudentTimetable(
    class_group_id?: number,
    day_of_week?: string,
): Promise<TimetableEntry[]> {
    return apiGet<TimetableEntry[]>(
        withQuery(
            `${TIMETABLE_BASE}/student/me`,
            {
                class_group_id,
                day_of_week,
            },
        ),
    );
}

export async function getParentChildTimetable(
    studentId: number,
    class_group_id?: number,
    day_of_week?: string,
): Promise<TimetableEntry[]> {
    return apiGet<TimetableEntry[]>(
        withQuery(
            `${TIMETABLE_BASE}/parent/child/${studentId}`,
            {
                class_group_id,
                day_of_week,
            },
        ),
    );
}

export async function listTimetableAssignments(
    filters?: TimetableAssignmentFilters,
): Promise<TimetableAssignment[]> {
    return apiGet<TimetableAssignment[]>(
        withQuery(
            `${TIMETABLE_BASE}/assignments`,
            filters,
        ),
    );
}

export async function createTimetableAssignment(
    payload: Partial<TimetableAssignment>,
): Promise<TimetableAssignment> {
    return apiPost<TimetableAssignment>(
        `${TIMETABLE_BASE}/assignments`,
        payload,
    );
}